rectools/models/candidate_ranking.py

```diff
@@ -0,0 +1,637 @@
+
+    def _get_user_features(
+        self, users: ExternalIds, dataset: Dataset, fold_info: tp.Optional[tp.Dict[str, tp.Any]]
+    ) -> pd.DataFrame:
+        return pd.DataFrame(columns=[Columns.User])
+
+    def _get_item_features(
+        self, items: ExternalIds, dataset: Dataset, fold_info: tp.Optional[tp.Dict[str, tp.Any]]
+    ) -> pd.DataFrame:
+        return pd.DataFrame(columns=[Columns.Item])
+
+    def _get_user_item_features(
+        self, useritem: pd.DataFrame, dataset: Dataset, fold_info: tp.Optional[tp.Dict[str, tp.Any]]
+    ) -> pd.DataFrame:
+        return pd.DataFrame(columns=Columns.UserItem)
+
+    def collect_features(
+        self, useritem: pd.DataFrame, dataset: Dataset, fold_info: tp.Optional[tp.Dict[str, tp.Any]]
+    ) -> pd.DataFrame:
+        """
+        Collect features for users-item pairs from any desired sources.
+
+        Parameters
+        ----------
+        useritem : pd.DataFrame
+            Candidates with score/rank features from first stage. Ids are either external or 1x internal
+        dataset : Dataset
+            Dataset will have either external -> 2x internal id maps to internal -> 2x internal
+        fold_info : tp.Optional[tp.Dict[str, tp.Any]]
+            Fold inofo from splitter can be used for adding time-based features
+
+        Returns
+        -------
+        pd.DataFrame
+            `useritem` dataframe enriched with features for users, items and useritem pairs
+        """
+
+        user_features = self._get_user_features(useritem[Columns.User].unique(), dataset, fold_info)
+        item_features = self._get_item_features(useritem[Columns.Item].unique(), dataset, fold_info)
+        useritem_features = self._get_user_item_features(useritem, dataset, fold_info)
+
+        res = useritem
+        res = res.merge(user_features, on=Columns.User, how="left")
+        res = res.merge(item_features, on=Columns.Item, how="left")
+        res = res.merge(useritem_features, on=Columns.UserItem, how="left")
+        return res
+
+
+@attr.s(auto_attribs=True)
+class NegativeSamplerBase:
+    def sample_negatives(self, train: pd.DataFrame) -> pd.DataFrame:
+        raise NotImplementedError()
+
+
+@attr.s(auto_attribs=True)
+class PerUserNegativeSampler(NegativeSamplerBase):
+    num_neg_samples: int = 3
+    random_state: tp.Optional[int] = None
+
+    def sample_negatives(self, train: pd.DataFrame) -> pd.DataFrame:
+        # train: user_id, item_id, scores, ranks, target(1/0)
+
+        negative_mask = train["target"] == 0
+        pos = train[~negative_mask]
+        neg = train[negative_mask]
+
+        # Some users might not have enough negatives for sampling
+        num_negatives = neg.groupby([Columns.User])[Columns.Item].count()
+        sampling_mask = train[Columns.User].isin(num_negatives[num_negatives > self.num_neg_samples].index)
+
+        neg_for_sample = train[sampling_mask & negative_mask]
+        neg = neg_for_sample.groupby([Columns.User], sort=False).apply(
```

Comment:
I see it's experimental, but I don't see TODO, so I'll just remind that `apply` is super super slow and we're not using it.  
I'm ok with TODO comment here if you don't want fixing it now

Reply:
A possible solution (maybe not the best one, but without using apply)

1. After calculating `num_negatives` calculate also the sampling ratio as `minimum(self.num_neg_samples / num_negatives, 1)`
2. Convert it to series with user_id as index and map it to the `neg` dataset
3. Add a column with random values in [0, 1)
4. Filter out rows with random values > sampling ratio

You also don't need a sampling mask here.  
And you can avoid splitting to pos and neg parts simply putting `1` as a sampling ratio for positives. This also means you don't need to do shuffling in the end. And I really don't like this shuffling, because:
1. It's a heavy operation
2. You have `sample` method in the `Sampler` class and it's quite not obvious that you shuffle the data inside

Reply:
Added TODO

Reply:
Add there a link to this conversation pls